// discord.js-commando/src/commands/util/prefix.js
const { stripIndents, oneLine } = require('common-tags');
const Command = require('../base');

module.exports = class PrefixCommand extends Command {
	constructor(client) {
		super(client, {
			name: 'prefix',
			group: 'util',
			memberName: 'prefix',
			description: 'Mostra ou define um prefixo.',
			format: '[prefix/"default"/"none"]',
			details: oneLine`
				Se não for dado nenhum prefixo, o atual será mostrado.
				Se o prefixo dado for *default* será restaurado o original.
				Se o prefixo for "none", o prefixo será retirado, e os comandos poderão apenas ser executados por menções.
				Apenas administradores podem alterar o prefixo.
			`,
			examples: ['prefix', 'prefix -', 'prefix omg!', 'prefix default', 'prefix none'],

			args: [
				{
					key: 'prefix',
					prompt: 'Qual o prefixo que pretende utilizar?',
					type: 'string',
					max: 15,
					default: ''
				}
			]
		});
	}

	async run(msg, args) {
		// Just output the prefix
	
		if(!args.prefix) {
			const prefix = msg.guild ? msg.guild.commandPrefix : this.client.commandPrefix;

			return msg.reply(stripIndents`
				${prefix ? `O prefixo é \`${prefix}\`.` : 'Não há prefixo.'}
				Para correr algum comando, use ${msg.anyUsage('comando')}.
			`);
		}

		// Check the user's permission before changing anything
		if(msg.guild) {
			if(!msg.member.hasPermission('ADMINISTRATOR') && !this.client.isOwner(msg.author)) {
				return msg.reply('Apenas o administrador pode alterar o prefixo.');
			}
		} else if(!this.client.isOwner(msg.author)) {
			return msg.reply('Apenas o dono do bot pode alterar o prefixo global.');
		}

		// Save the prefix
		const lowercase = args.prefix.toLowerCase();
		const prefix = lowercase === 'none' ? '' : args.prefix;
		let response;
		if(lowercase === 'default') {
			if(msg.guild) msg.guild.commandPrefix = null; else this.client.commandPrefix = null;
			const current = this.client.commandPrefix ? `\`${this.client.commandPrefix}\`` : 'no prefix';
			response = `Prefixo restaurado para a origem (atualmente ${current}).`;
		} else {
			if(msg.guild) msg.guild.commandPrefix = prefix; else this.client.commandPrefix = prefix;
			response = prefix ? `Prefixo alterado para \`${args.prefix}\`.` : 'Prefixo removido por completo.';
		}

		msg.reply(`${response} Para correr qualque comando, use ${msg.anyUsage('comando')}.`);
		return null;
	}
};
