// doyle-music/index.js
const Discord = require("discord.js");
const youtube = require("simple-youtube-api");
const handleVideo = require("./modules/MusicHandling.js");

class Music {

  /**
    * @param {Client} client The Discord.JS client
    * @param {String} token Youtube API Token
    * @param  {Object} [options = {}] The options to pass
    * @example
    * // an example of initiating the music client
    * const MusicApi = require("doyle-music");
    * const music = new api(client, "YT API KEY", { prefix: "-", anyoneCanSkip: true, autoHandle: false, autoLeaveTime: 30000 });
    */

  constructor(client, token, options) {
    if (!client) throw new Error("You must pass a discord.js client!");
    if (typeof client !== "object") throw new Error("Client must be an actual client!");
    this.client = client;

    this.client.embed = require("./modules/Embeds.js");

    if (!token.length) throw new Error("YT API Token is required!");
    if (typeof token !== "string") throw new Error("Token must be a string!");
    this.token = token;
    

    this.prefix = (options && options.prefix) || "!";
    if (typeof this.prefix !== "string") throw new Error("Prefix must be a string!");


    this.client.anyoneCanSkip = (options && options.anyoneCanSkip) || false;
    if (typeof this.client.anyoneCanSkip !== "boolean") throw new Error("anyoneCanSkip must be a boolean!");


    this.autoHandle = (options && options.autoHandle) || false;
    if (typeof this.autoHandle !== "boolean") throw new Error("autoHandle must be a boolean!");


    this.client.autoLeaveTime = (options && options.autoLeaveTime) || 30000;
    if (typeof this.client.autoLeaveTime !== "number") throw new Error("autoLeaveTime must be a number!");


    if (options && typeof options !== "object") throw new Error("Options must be an object!");

    this.client.ytapi = new youtube(token);
    this.client.playlists = new Discord.Collection();

    this.client.on("ready", () => {
      console.log("Music BOT is on!");
    });

    this.client.on("message", async message => {
      if (message.author.bot) return;
      if (!this.autoHandle) return;
      if (!message.content.startsWith(this.prefix)) return;
      let args = message.content.split(" ");
      const command = args[0].slice(this.prefix.length);
      args = args.join(" ").slice(command.length + this.prefix.length).trim().split(" ");

      if (command === "play") {
        this.play(this.client, message, args);
      }

      if (command === "skip") {
        this.skip(this.client, message);
      }

      if (command === "loop" || command === "unloop") {
        this.loop(this.client, message);
      }

      if (command === "np" || command === "nowplaying" || command === "currentsong") {
        this.np(this.client, message);
      }

      if (command === "queue" || command === "playlist") {
        this.queue(client, message);
      }

      if (command === "pause") {
        this.pause(client, message);
      }

      if (command === "resume") {
        this.resume(client, message);
      } 

      if (command === "search") {
        this.search(client, message, args);
      }

      if (command === "stop") {
        this.stop(client, message);
      }
      if (command === "volume") {
        this.volume(client, message, args);
      }
    });
  }

  /**
   * @param {DiscordClient} client The discord.js client
   * @param {DiscordMessage} message The message sent, a discord.js message object
   * @param {Array} [args = []] The arguments, look at the example for more information
   * @description This will handle all of the stuff to play a song using a video id, url or name
   */

  async play(client, message, args = []) {
    const url = args[0] ? args[0].replace(/<(.+)>/g, "$1") : "";
    const voiceChannel = message.member.voiceChannel;
    if (!voiceChannel) return client.embed("noVoiceChannel", message);
    if (!args[0]) return client.embed("noArgs", message);
    const permissions = voiceChannel.permissionsFor(client.user);
    if (!permissions.has("CONNECT")) return client.embed("noPerms-CONNECT", message);
    if (!permissions.has("SPEAK")) return client.embed("noPerms-SPEAK", message);
    if (url.match(/^https?:\/\/(www.youtube.com|youtube.com)\/playlist(.*)$/)) {
      const playlist = await client.ytapi.getPlaylist(url);
      const videos = await playlist.getVideos();
      for (const video of Object.values(videos)) {
        const video2 = await client.ytapi.getVideoByID(video.id); // eslint-disable-line no-await-in-loop
        await handleVideo(video2, message, voiceChannel, true); // eslint-disable-line no-await-in-loop
      }
      const embed = new Discord.MessageEmbed()
        .setAuthor("Playlist")
        .setDescription(`✅ Playlist: **${playlist.title}** has been added to the queue!`)
        .setColor(0x00AE86);
      message.channel.send({ embed: embed });
    } else {
      let video;
      try {
        video = await client.ytapi.getVideo(url);
      } catch (error) {
        const videos = await client.ytapi.searchVideos(args.join(" "), 1);
        video = await client.ytapi.getVideoByID(videos[0].id);   
      }
      return handleVideo(video, message, voiceChannel);
    }
  }

  /**
   * @param {DiscordClient} client The discord.js client
   * @param {DiscordMessage} message The message sent, a discord.js message object
   * @description If a song is playing, this will skip it
   */


  skip(client, message) {
    if (!client.anyoneCanSkip && message.member.hasPermission("ADMINISTRATOR")) return client.embed("noPerms-ADMINISTRATOR", message); 
    const voiceChannel = message.member.voiceChannel;
    if (!voiceChannel) return client.embed("noVoiceChannel", message);
    if (!client.playlists.has(message.guild.id)) return client.embed("emptyQueue", message);
    const thisPlaylist = client.playlists.get(message.guild.id);
    thisPlaylist.loop = false;
    thisPlaylist.connection.dispatcher.end("skip");
    return client.embed("skipped", message);
  }

  /**
   * @param {DiscordClient} client The discord.js client
   * @param {DiscordMessage} message The message sent, a discord.js message object
   * @description If there is a song playing, this will loop it
   */

  loop(client, message) {
    const voiceChannel = message.member.voiceChannel;
    const thisPlaylist = client.playlists.get(message.guild.id);
    if (!voiceChannel) return client.embed("noVoiceChannel", message);
    if (!client.playlists.has(message.guild.id)) return client.embed("emptyQueue", message);
    if (thisPlaylist.loop) {
      thisPlaylist.loop = false;
      return client.embed("unloopedEmbed", message);
    } else {
      thisPlaylist.loop = true;
      return client.embed("loopedEmbed", message);
    }
  }

  /**
   * @param {DiscordClient} client The discord.js client
   * @param {DiscordMessage} message The message sent, a discord.js message object
   * @description This will display the current playing song
   */


  np(client, message) {
    const voiceChannel = message.member.voiceChannel;
    if (!voiceChannel) return client.embed("noVoiceChannel", message);
    if (!client.playlists.has(message.guild.id)) return client.embed("emptyQueue", message);
    return client.embed("nowPlaying", message);
  }

  /**
   * @param {DiscordClient} client The discord.js client
   * @param {DiscordMessage} message The message sent, a discord.js message object
   * @param {Array} [args = []] The arguments, look at the example for more information
   * @description If there is a current song, this will pause it
   */


  pause(client, message) {
    const voiceChannel = message.member.voiceChannel;
    if (!voiceChannel) return client.embed("noVoiceChannel", message);
    if (!client.playlists.has(message.guild.id)) return client.embed("emptyQueue", message);    
    const thisPlaylist = client.playlists.get(message.guild.id);
    if (!thisPlaylist.playing) return client.embed("alreadyPaused", message);   
    thisPlaylist.playing = false;
    thisPlaylist.connection.dispatcher.pause();
    return client.embed("paused", message);  
  }

  /**
   * @param {DiscordClient} client The discord.js client
   * @param {DiscordMessage} message The message sent, a discord.js message object
   * @description This will display the current playing songs
   */


  queue(client, message) {
    const voiceChannel = message.member.voiceChannel;
    if (!voiceChannel) return client.embed("noVoiceChannel", message);
    if (!client.playlists.has(message.guild.id)) return client.embed("emptyQueue", message);
    return client.embed("queueEmbed", message);
  }

  /**
   * @param {DiscordClient} client The discord.js client
   * @param {DiscordMessage} message The message sent, a discord.js message object
   * @description If the current song is paused, it will resume it
   */


  resume(client, message) {
    const voiceChannel = message.member.voiceChannel;
    if (!voiceChannel) return client.embed("noVoiceChannel", message);
    if (!client.playlists.has(message.guild.id)) return client.embed("emptyQueue", message);
    const thisPlaylist = client.playlists.get(message.guild.id);
    if (thisPlaylist.playing) return client.embed("alreadyResumed", message);
    thisPlaylist.playing = true;
    thisPlaylist.connection.dispatcher.resume();
    return client.embed("resumed", message);
  }

  /**
   * @param {DiscordClient} client The discord.js client
   * @param {DiscordMessage} message The message sent, a discord.js message object
   * @param {Array} [args = []] The arguments, look at the example for more information
   * @description This will display a list of 10 songs based on searches and allow you to choose one
   */

  async search(client, message, args) {
    const voiceChannel = message.member.voiceChannel;
    const permissions = voiceChannel.permissionsFor(client.user);
    if (!voiceChannel) return client.embed("noVoiceChannel", message);
    if (!args[0]) return client.embed("noArgs", message);
    if (!permissions.has("CONNECT")) return client.embed("noPerms-CONNECT", message);
    if (!permissions.has("SPEAK")) return client.embed("noPerms-SPEAK");
    let video;
    try {
      const videos = await youtube.searchVideos(args.join(" "), 10);
      let index = 0;
      const embed = new Discord.MessageEmbed()
        .setAuthor("Song Selection")
        .setDescription(`${videos.map(video2 => `**${++index} -** ${video2.title}`).join("\n")}`)
        .setFooter("Please provide a value to select one of the search results ranging from 1-10.")
        .setColor(0x00AE86);
      message.channel.send({ embed: embed });
      const response = await message.channel.awaitMessages(msg2 => (msg2.content > 0 && msg2.content < 11) && msg2.author.id === message.author.id, {
        max: 1,
        time: 10000,
        errors: ["time"]
      });
      if (!response) return client.embed("invalidEntry", message);
      const videoIndex = parseInt(response.first().content);
      video = await youtube.getVideoByID(videos[videoIndex - 1].id);
    } catch (err) {
      return client.embed("noSearchResults", message);
    }
    return handleVideo(video, message, voiceChannel);
  }

  /**
   * @param {DiscordClient} client The discord.js client
   * @param {DiscordMessage} message The message sent, a discord.js message object
   * @description If there is a queue, this will clear it and leave the voice channel
   */


  stop(client, message) {
    if (message.member.hasPermission("ADMINISTRATOR")) {
      const voiceChannel = message.member.voiceChannel;
      if (!voiceChannel) return client.embed("noVoiceChannel", message);
      if (!client.playlists.has(message.guild.id)) return client.embed("emptyQueue", message);
      const thisPlaylist = client.playlists.get(message.guild.id);
      thisPlaylist.songs = [];
      thisPlaylist.connection.dispatcher.end();
      return client.embed("stopped", message);
    } else {
      return client.embed("noPerms-ADMINISTRATOR", message);
    }
  }

  /**
   * @param {DiscordClient} client The discord.js client
   * @param {DiscordMessage} message The message sent, a discord.js message object
   * @param {Array} [args = []] The arguments, look at the example for more information
   * @description This will set the volume of the current queue
   */


  volume(client, message, args) {
    if (message.member.hasPermission("ADMINISTRATOR")) {
      const voiceChannel = message.member.voiceChannel;
      if (!voiceChannel) return client.embed("noVoiceChannel", message);
      if (!client.playlists.has(message.guild.id)) return client.embed("emptyQueue", message);
      if (!args[0]) return client.embed("currentVolume", message);
      if (Number(args[0]) < 0 || Number(args[0]) > 100) return client.embed("errorVolume", message);
      message.guild.voiceConnection.volume = Number(args[0]) / 100;
      client.playlists.get(message.guild.id).volume = Number(args[0]);
      client.playlists.get(message.guild.id).connection.dispatcher.setVolumeLogarithmic(Number(args[0]) / 100);
      return client.embed("volumeSet", message, args);  
    } else {
      return client.embed("noPerms-ADMINISTRATOR", message);
    }
  }
}

module.exports = Music;
